src/rt/aaA.d

```diff
@@ -142,16 +142,36 @@ private:
     {
         auto obuckets = buckets;
         buckets = allocBuckets(ndim);
+        auto newFirstUsed = cast(uint) ndim;
 
-        foreach (ref b; obuckets)
+        foreach (ref b; obuckets[firstUsed .. $])
             if (b.filled)
-                *findSlotInsert(b.hash) = b;
+            {
+                auto x = findSlotInsert(b.hash);
+                newFirstUsed = min(newFirstUsed, cast(uint) (x - buckets.ptr));
```

Comment:
This change needs to pass our AA benchmark suite, AA resizing is very performance critical and you're adding a bunch of instructions in the hot loop.
My gut feeling would be that it's cheaper to compute the firstUsed cache after repopulating the new table, and even cheaper to do it lazily next time it's used.